internal/api/phone.go

```diff
@@ -104,3 +104,16 @@ func (a *API) sendPhoneConfirmation(ctx context.Context, tx *storage.Connection,
 
 	return messageID, errors.Wrap(tx.UpdateOnly(user, includeFields...), "Database error updating user for confirmation")
 }
+
+func parseSmsTemplate(smsTemplate, otp string) (string, error) {
+	if smsTemplate == "" {
+		return fmt.Sprintf(defaultSmsMessage, otp), nil
+	}
+	var message bytes.Buffer
+	if err := template.Must(template.New(defaultTemplateName).Parse(smsTemplate)).Execute(&message, struct {
+		Code string
+	}{Code: otp}); err != nil {
+		return "", err
+	}
+	return message.String(), nil
+}
```

Comment:
Parsing the template is a bit slow and processor heavy. Maybe we can move it to the config, so it's done only once, and we can do something like `a.db.Sms.Template.Execute(...)` each time instead?

Reply:
hmm isn't the template fetched on demand for our use case? if we were to parse it in the config and the template is changed, the error won't be surfaced if an incorrect template string is used.

Reply:
Actually yea, the otp is only generated when the message is sent so not sure how best to obtain otp and  parse the template in config

Reply:
i would prefer parsing it on the UI / database level before the config is being set in gotrue - i think it's better for the validation to be handled externally and assume that the config loaded into gotrue is always correct

Reply:
Well that's not always the case, if you're self hosting. So validation doesn't hurt.

Reply:
> hmm isn't the template fetched on demand for our use case? if we were to parse it in the config and the template is changed, the error won't be surfaced if an incorrect template string is used.

What do you mean on demand? It's read directly from the env config like so:

```go
	if config.Sms.Template == "" {
		message = fmt.Sprintf(defaultSmsMessage, otp)
	} else {
		message = strings.Replace(config.Sms.Template, "{{ .Code }}", otp, -1)
```

I don't see how that template string contained in `config.Sms.Template` is dynamic?

Reply:
Yeah my bad i got this confused with the `GOTRUE_MAILER_URLPATHS_XXX` env vars which is what we use to load the email template. It's fine to do the validation when the config is loaded for the `GOTRUE_MAILER_TEMPLATES_XXX` and `GOTRUE_SMS_TEMPLATE` env vars - just wanted to point out that it's not gonna be reflected on the hosted gotrue service because that uses the URL paths, which is fetched every time an email needs to be sent out.

Reply:
Thanks for the input everyone! Will move the parsing logic to the config